File: src_lua/ReplicatedStorage/ClientSource/Client/CatController/Components/Others/CatRenderer.lua
local CatRenderer = {}

-- External dependencies
local CatService

-- Internal state
local CatRenderer = {
	CatVisuals = {},
	CatModels = {},
	MoodIndicators = {},
	ParticleEffects = {}
}

function CatRenderer:SpawnCatVisual(catId, catData)
	return self:CreateCatVisual(catId, catData)
end

function CatRenderer:CreateCatVisual(catId, catData)
	-- Clone the Petra cat model from workspace
	local petraModel = game.Workspace.Models:FindFirstChild("Petra")
	if not petraModel then
		warn("Petra cat model not found in Workspace.Models")
		return nil
	end
	
	-- Clone the model for this cat
	local catVisual = petraModel:Clone()
	catVisual.Name = "Cat_" .. catId
	
	-- Configure humanoid for cat behavior
	local humanoid = catVisual:FindFirstChildOfClass("Humanoid")
	if humanoid then
		humanoid.WalkSpeed = catData.profile.physical.movementSpeed
		humanoid.JumpPower = catData.profile.physical.jumpHeight
		
		-- Configure humanoid for cat-like behavior
		humanoid.AutoRotate = true
		humanoid.AutoJumpEnabled = false
		humanoid.DisplayDistanceType = Enum.HumanoidDisplayDistanceType.None
		
		-- Set up animations for cat behaviors
		self:SetupCatAnimations(catId, humanoid)
	end
	
	-- Position the cat in the world
	if catVisual.PrimaryPart then
		catVisual:SetPrimaryPartCFrame(CFrame.new(catData.currentState.position))
	else
		-- Fallback positioning
		local root = catVisual:FindFirstChild("Root") or catVisual:FindFirstChild("Torso")
		if root then
			root.CFrame = CFrame.new(catData.currentState.position)
		end
	end
	
	catVisual.Parent = workspace
	
	-- Store visual reference
	CatRenderer.CatVisuals[catId] = catVisual
	CatRenderer.CatModels[catId] = catData
	
	-- Create mood indicator
	CatRenderer:CreateMoodIndicator(catId, catData)
	
	print("Created visual for cat:", catId, "using Petra model")
	
	return catVisual
end

function CatRenderer:DestroyCatVisual(catId)
	if CatRenderer.CatVisuals[catId] then
		CatRenderer.CatVisuals[catId]:Destroy()
		CatRenderer.CatVisuals[catId] = nil
		CatRenderer.CatModels[catId] = nil
		
		-- Clean up mood indicator
		if CatRenderer.MoodIndicators[catId] then
			CatRenderer.MoodIndicators[catId]:Destroy()
			CatRenderer.MoodIndicators[catId] = nil
		end
		
		print("Destroyed visual for cat:", catId)
	end
end

function CatRenderer:UpdateCatVisual(catId, catData)
	local catVisual = CatRenderer.CatVisuals[catId]
	if not catVisual then return end
	
	-- Update humanoid properties and movement
	local humanoid = catVisual:FindFirstChildOfClass("Humanoid")
	if humanoid then
		humanoid.WalkSpeed = catData.profile.physical.movementSpeed
		humanoid.JumpPower = catData.profile.physical.jumpHeight
		
		-- Make the cat actually move to the target position
		if catData.behaviorState and catData.behaviorState.isMoving then
			local currentPos = catVisual.PrimaryPart and catVisual.PrimaryPart.Position or Vector3.new(0, 0, 0)
			local targetPos = catData.currentState.position
			
			-- Only move if we're not already at the target
			if (targetPos - currentPos).Magnitude > 1 then
				local direction = (targetPos - currentPos).Unit
				humanoid:MoveTo(targetPos)
				print("🐱 [CatRenderer] Cat", catId, "moving to:", targetPos)
			else
				humanoid:MoveTo(currentPos) -- Stop moving
			end
		else
			-- Stop moving if not supposed to be moving
			humanoid:MoveTo(catVisual.PrimaryPart and catVisual.PrimaryPart.Position or Vector3.new(0, 0, 0))
		end
	end
	
	-- Update stored data
	CatRenderer.CatModels[catId] = catData
	
	-- Update mood indicator
	CatRenderer:UpdateMoodIndicator(catId, catData.moodState)
end

function CatRenderer:CreateMoodIndicator(catId, catData)
	-- Create a billboard GUI to show mood
	local moodIndicator = Instance.new("BillboardGui")
	moodIndicator.Name = "MoodIndicator"
	moodIndicator.Size = UDim2.new(4, 0, 1, 0)
	moodIndicator.StudsOffset = Vector3.new(0, 3, 0)
	moodIndicator.AlwaysOnTop = true
	
	local moodFrame = Instance.new("Frame")
	moodFrame.Size = UDim2.new(1, 0, 1, 0)
	moodFrame.BackgroundTransparency = 1
	moodFrame.Parent = moodIndicator
	
	local moodLabel = Instance.new("TextLabel")
	moodLabel.Size = UDim2.new(1, 0, 1, 0)
	moodLabel.BackgroundTransparency = 1
	moodLabel.Text = catData.moodState.currentMood
	moodLabel.TextColor3 = CatRenderer:GetMoodColor(catData.moodState.currentMood)
	moodLabel.TextScaled = true
	moodLabel.Font = Enum.Font.GothamBold
	moodLabel.Parent = moodFrame
	
	-- Attach to cat's head
	local catVisual = CatRenderer.CatVisuals[catId]
	if catVisual then
		local head = catVisual:FindFirstChild("Head")
		if head then
			moodIndicator.Adornee = head
			moodIndicator.Parent = head
		else
			-- Fallback: attach to torso
			local torso = catVisual:FindFirstChild("Torso")
			if torso then
				moodIndicator.Adornee = torso
				moodIndicator.Parent = torso
			end
		end
	end
	
	CatRenderer.MoodIndicators[catId] = moodIndicator
end

function CatRenderer:UpdateMoodIndicator(catId, moodState)
	local moodIndicator = CatRenderer.MoodIndicators[catId]
	if not moodIndicator then return end
	
	local moodLabel = moodIndicator:FindFirstChild("Frame"):FindFirstChild("TextLabel")
	if moodLabel then
		moodLabel.Text = moodState.currentMood
		moodLabel.TextColor3 = CatRenderer:GetMoodColor(moodState.currentMood)
	end
end

function CatRenderer:GetMoodColor(moodType)
	local moodColors = {
		Happy = Color3.fromRGB(76, 175, 80),    -- Green
		Curious = Color3.fromRGB(33, 150, 243), -- Blue
		Annoyed = Color3.fromRGB(255, 152, 0),  -- Orange
		Hungry = Color3.fromRGB(244, 67, 54),   -- Red
		Tired = Color3.fromRGB(156, 39, 176),   -- Purple
		Afraid = Color3.fromRGB(121, 85, 72),   -- Brown
		Playful = Color3.fromRGB(255, 193, 7)   -- Yellow
	}
	
	return moodColors[moodType] or Color3.fromRGB(189, 189, 189) -- Default gray
end

function CatRenderer:PlayParticleEffect(catId, effectType)
	-- Create particle effects for various cat behaviors
	local catVisual = CatRenderer.CatVisuals[catId]
	if not catVisual then return end
	
	local particleEffect = Instance.new("ParticleEmitter")
	particleEffect.Name = effectType .. "Effect"
	
	-- Configure based on effect type
	if effectType == "Eat" then
		particleEffect.Texture = "rbxasset://textures/particles/sparkles_main.dds"
		particleEffect.Lifetime = NumberRange.new(0.5, 1.5)
		particleEffect.Rate = 20
		particleEffect.SpreadAngle = Vector2.new(45, 45)
		particleEffect.Color = ColorSequence.new(Color3.fromRGB(255, 255, 200))
	elseif effectType == "Play" then
		particleEffect.Texture = "rbxasset://textures/particles/sparkles_main.dds"
		particleEffect.Lifetime = NumberRange.new(0.3, 0.8)
		particleEffect.Rate = 30
		particleEffect.SpreadAngle = Vector2.new(90, 90)
		particleEffect.Color = ColorSequence.new(Color3.fromRGB(255, 200, 100))
	end
	
	-- Attach to cat's head or torso
	local head = catVisual:FindFirstChild("Head")
	local torso = catVisual:FindFirstChild("Torso")
	local attachTo = head or torso
	
	if attachTo then
		particleEffect.Parent = attachTo
		
		-- Auto-destroy after duration
		task.delay(3, function()
			if particleEffect and particleEffect.Parent then
				particleEffect:Destroy()
			end
		end)
	end
	
	CatRenderer.ParticleEffects[catId] = particleEffect
end

function CatRenderer:SetupCatAnimations(catId, humanoid)
	-- Set up animation tracks for various cat behaviors
	-- These will be played by the AnimationHandler component
	
	-- Load default animations (placeholder IDs - replace with actual cat animations)
	local animator = humanoid:FindFirstChildOfClass("Animator")
	if not animator then
		animator = Instance.new("Animator")
		animator.Parent = humanoid
	end
	
	-- Store animation tracks for later use
	CatRenderer.AnimationTracks = CatRenderer.AnimationTracks or {}
	CatRenderer.AnimationTracks[catId] = {
		Walk = nil,
		Run = nil,
		Idle = nil,
		Sit = nil,
		Sleep = nil,
		Play = nil,
		Eat = nil,
		Groom = nil
	}
	
	print("Set up animations for cat:", catId)
end

function CatRenderer:GetCatVisual(catId)
	return CatRenderer.CatVisuals[catId]
end

function CatRenderer:GetAllCatVisuals()
	return CatRenderer.CatVisuals
end

function CatRenderer:IsCatVisible(catId)
	local catVisual = CatRenderer.CatVisuals[catId]
	return catVisual and catVisual.Parent ~= nil
end

function CatRenderer:SetCatVisibility(catId, visible)
	local catVisual = CatRenderer.CatVisuals[catId]
	if catVisual then
		catVisual.Parent = visible and workspace or nil
		
		-- Also hide/show mood indicator
		local moodIndicator = CatRenderer.MoodIndicators[catId]
		if moodIndicator then
			moodIndicator.Enabled = visible
		end
	end
end

-- Component initialization
function CatRenderer.Init()
	-- Get reference to parent CatController
	local CatController = script.Parent.Parent.Parent
	
	print("CatRenderer component initialized")
end

function CatRenderer:CullDistantCats()
	-- Cull distant cats based on performance settings
	local player = game.Players.LocalPlayer
	local character = player.Character
	
	if not character or not character:FindFirstChild("HumanoidRootPart") then
		return
	end
	
	local playerPosition = character.HumanoidRootPart.Position
	
	-- Get all active cat visuals
	for catId, catVisual in pairs(CatRenderer.ActiveCatVisuals or {}) do
		if catVisual and catVisual.PrimaryPart then
			local distance = (catVisual.PrimaryPart.Position - playerPosition).Magnitude
			
			-- Simple distance-based culling
			if distance > 200 then -- Max render distance
				catVisual.Parent = nil -- Hide the cat
			else
				catVisual.Parent = workspace -- Show the cat
			end
		end
	end
	
	print("Culled distant cats based on player position")
end

function CatRenderer.Start()
	print("CatRenderer component started")
end

return CatRenderer

File: src/client/cat-renderer.ts
import { Workspace, Players } from "@rbxts/services";
import { CatData, MoodState } from "shared/cat-types";
import { AnimationHandler } from "./animation-handler";
import { PhysicsService } from "@rbxts/services";

export class CatRenderer {
    private static catVisuals = new Map<string, Model>();
    private static moodIndicators = new Map<string, BillboardGui>();
    private static previousActions = new Map<string, string>();

    public static CreateCatVisual(catId: string, catData: CatData): Model | undefined {
        const modelsFolder = Workspace.FindFirstChild("Models");
        if (!modelsFolder) {
            warn("Models folder not found in Workspace");
            return;
        }

        const petraModel = modelsFolder.FindFirstChild("Petra") as Model;
        if (!petraModel) {
            warn("Petra cat model not found in Workspace.Models");
            return;
        }

        const catVisual = petraModel.Clone();
        catVisual.Name = `Cat_${catId}`;

        const humanoid = catVisual.FindFirstChildOfClass("Humanoid");
        if (humanoid) {
            humanoid.WalkSpeed = catData.profile.physical.movementSpeed;
            humanoid.JumpPower = catData.profile.physical.jumpHeight;
            humanoid.AutoRotate = true;
            humanoid.AutoJumpEnabled = false;
            humanoid.DisplayDistanceType = Enum.HumanoidDisplayDistanceType.None;
        }

        if (catVisual.PrimaryPart) {
            catVisual.SetPrimaryPartCFrame(new CFrame(catData.currentState.position));
        } else {
            const root = catVisual.FindFirstChild("Root") as BasePart || catVisual.FindFirstChild("Torso") as BasePart;
            if (root) {
                root.CFrame = new CFrame(catData.currentState.position);
            }
        }

        for (const d of catVisual.GetDescendants()) {
            if (d.IsA("BasePart")) {
                d.Anchored = false;
                d.CanCollide = true;
                d.Massless = false;
                // PhysicsService.SetPartCollisionGroup(d, "Cats");
            }
        }

        catVisual.Parent = Workspace;
        this.catVisuals.set(catId, catVisual);

        this.CreateMoodIndicator(catId, catData);

        return catVisual;
    }

    public static RemoveCatVisual(catId: string) {
        const visual = this.catVisuals.get(catId);
        if (visual) {
            AnimationHandler.StopAnimation(catId);
            
            // Stop purring sound if playing
            this.StopPurringSound(catId);
            
            visual.Destroy();
            this.catVisuals.delete(catId);
        }

        const indicator = this.moodIndicators.get(catId);
        if (indicator) {
            indicator.Destroy();
            this.moodIndicators.delete(catId);
        }

        // Clean up previous action tracking
        this.previousActions.delete(catId);
    }

    public static UpdateCatVisual(catId: string, catData: CatData) {
        const visual = this.catVisuals.get(catId);
        if (!visual) return;

        const humanoid = visual.FindFirstChildOfClass("Humanoid");
        if (humanoid) {
            humanoid.WalkSpeed = catData.profile.physical.movementSpeed;
            humanoid.JumpPower = catData.profile.physical.jumpHeight;

            const action = catData.behaviorState.currentAction;
            const targetPos = catData.behaviorState.targetPosition;
            const currentPos = visual.PrimaryPart?.Position || new Vector3();

            // Stop purring if action changed from Purr to something else
            const previousAction = this.previousActions.get(catId);
            if (previousAction === "Purr" && action !== "Purr") {
                this.StopPurringSound(catId);
            }
            this.previousActions.set(catId, action);

            if (catData.behaviorState.isMoving && targetPos) {
                if (targetPos.sub(currentPos).Magnitude > 0.5) {
                    humanoid.MoveTo(targetPos);
                    AnimationHandler.PlayAnimation(catId, "Walk", humanoid);
                } else {
                    humanoid.MoveTo(currentPos);
                    AnimationHandler.PlayAnimation(catId, "Idle", humanoid);
                }
            } else {
                humanoid.MoveTo(currentPos);
                AnimationHandler.PlayAnimation(catId, action || "Idle", humanoid);

                // Special handling for Purr: look at the player who petted
                if (action === "Purr") {
                    const actionData = catData.behaviorState.actionData as { reactingToPlayerId?: number } | undefined;
                    if (actionData?.reactingToPlayerId) {
                        const player = Players.GetPlayerByUserId(actionData.reactingToPlayerId);
                        const char = player?.Character;
                        const playerHRP = char?.FindFirstChild("HumanoidRootPart") as Part;
                        if (playerHRP) {
                            const playerPos = playerHRP.Position;
                            const lookVector = playerPos.sub(currentPos).Unit;
                            const flatLookVector = new Vector3(lookVector.X, 0, lookVector.Z).Unit;
                            if (flatLookVector.Magnitude > 0) {
                                visual.SetPrimaryPartCFrame(CFrame.lookAt(currentPos, currentPos.add(flatLookVector)));
                            }
                            // Play purring sound effect
                            this.PlayPurringSound(catId, visual);
                        }
                    } else if (targetPos) {
                        // Fallback to targetPos if no player data
                        const lookVector = targetPos.sub(currentPos).Unit;
                        const flatLookVector = new Vector3(lookVector.X, 0, lookVector.Z).Unit;
                        if (flatLookVector.Magnitude > 0) {
                            visual.SetPrimaryPartCFrame(CFrame.lookAt(currentPos, currentPos.add(flatLookVector)));
                        }
                        this.PlayPurringSound(catId, visual);
                    }
                } else {
                    // Rotation for LookAt/Follow/Meow/RollOver
                    if ((action === "LookAt" || action === "Follow" || action === "Meow" || action === "RollOver") && targetPos) {
                        const lookVector = targetPos.sub(currentPos).Unit;
                        const flatLookVector = new Vector3(lookVector.X, 0, lookVector.Z).Unit;
                        if (flatLookVector.Magnitude > 0) {
                            visual.SetPrimaryPartCFrame(CFrame.lookAt(currentPos, currentPos.add(flatLookVector)));
                        }
                    }
                }
            }

            // Special Meow visual
            if (action === "Meow") {
                this.ShowTemporaryMoodText(catId, "Meow!", Color3.fromRGB(255, 255, 255));
            }

            // Special Purr visual
            if (action === "Purr") {
                this.ShowTemporaryMoodText(catId, "Purr...", Color3.fromRGB(255, 200, 255));
            }
        }

        this.UpdateMoodIndicator(catId, catData.moodState);
        this.UpdateHoldingState(catId, visual, catData);
    }

    private static UpdateHoldingState(catId: string, visual: Model, catData: CatData) {
        const heldById = catData.behaviorState.heldByPlayerId;
        const currentWeld = visual.FindFirstChild("HoldingWeld") as Weld;

        if (heldById !== undefined) {
            const player = Players.GetPlayerByUserId(heldById);
            const character = player?.Character;
            const targetPart = character?.FindFirstChild("RightHand") as BasePart
                || character?.FindFirstChild("Right Arm") as BasePart
                || character?.FindFirstChild("HumanoidRootPart") as BasePart;

            if (targetPart && !currentWeld) {
                const weld = new Instance("Weld");
                weld.Name = "HoldingWeld";
                weld.Part0 = targetPart;
                weld.Part1 = visual.PrimaryPart || visual.FindFirstChildOfClass("BasePart");
                weld.C0 = new CFrame(new Vector3(0, -0.5, -1)); // Position offset
                weld.Parent = visual;

                const humanoid = visual.FindFirstChildOfClass("Humanoid");
                if (humanoid) {
                    humanoid.PlatformStand = true;
                }
            }
        } else if (currentWeld) {
            currentWeld.Destroy();
            const humanoid = visual.FindFirstChildOfClass("Humanoid");
            if (humanoid) {
                humanoid.PlatformStand = false;
            }
        }
    }

    private static CreateMoodIndicator(catId: string, catData: CatData) {
        const visual = this.catVisuals.get(catId);
        if (!visual) return;

        const indicator = new Instance("BillboardGui");
        indicator.Name = "MoodIndicator";
        indicator.Size = new UDim2(4, 0, 1, 0);
        indicator.StudsOffset = new Vector3(0, 3, 0);
        indicator.AlwaysOnTop = true;

        const frame = new Instance("Frame");
        frame.Size = new UDim2(1, 0, 1, 0);
        frame.BackgroundTransparency = 1;
        frame.Parent = indicator;

        const label = new Instance("TextLabel");
        label.Size = new UDim2(1, 0, 1, 0);
        label.BackgroundTransparency = 1;
        label.Text = catData.moodState.currentMood;
        label.TextColor3 = this.GetMoodColor(catData.moodState.currentMood);
        label.TextScaled = true;
        label.Font = Enum.Font.GothamBold;
        label.Parent = frame;

        const head = visual.FindFirstChild("Head") as BasePart || visual.FindFirstChild("Torso") as BasePart;
        if (head) {
            indicator.Adornee = head;
            indicator.Parent = head;
        }

        this.moodIndicators.set(catId, indicator);
    }

    private static UpdateMoodIndicator(catId: string, moodState: MoodState) {
        const indicator = this.moodIndicators.get(catId);
        if (!indicator) return;

        const label = indicator.FindFirstChild("Frame")?.FindFirstChild("TextLabel") as TextLabel;
        if (label && label.GetAttribute("IsTemporary") !== true) {
            label.Text = moodState.currentMood;
            label.TextColor3 = this.GetMoodColor(moodState.currentMood);
        }
    }

    private static ShowTemporaryMoodText(catId: string, text: string, color: Color3) {
        const indicator = this.moodIndicators.get(catId);
        if (!indicator) return;

        const label = indicator.FindFirstChild("Frame")?.FindFirstChild("TextLabel") as TextLabel;
        if (label) {
            label.Text = text;
            label.TextColor3 = color;
            label.SetAttribute("IsTemporary", true);

            task.delay(1.5, () => {
                if (label.Parent) {
                    label.SetAttribute("IsTemporary", false);
                }
            });
        }
    }

    private static GetMoodColor(moodType: string): Color3 {
        const colors: Record<string, Color3> = {
            Happy: Color3.fromRGB(76, 175, 80),
            Curious: Color3.fromRGB(33, 150, 243),
            Annoyed: Color3.fromRGB(255, 152, 0),
            Hungry: Color3.fromRGB(244, 67, 54),
            Tired: Color3.fromRGB(156, 39, 176),
            Afraid: Color3.fromRGB(121, 85, 72),
            Playful: Color3.fromRGB(255, 193, 7),
        };
        return colors[moodType] || Color3.fromRGB(189, 189, 189);
    }

    private static purringSounds = new Map<string, Sound>();

    private static StopPurringSound(catId: string) {
        const sound = this.purringSounds.get(catId);
        if (sound && sound.Parent) {
            sound.Stop();
        }
        this.purringSounds.delete(catId);
    }

    private static PlayPurringSound(catId: string, visual: Model) {
        // Only play sound if not already playing
        if (this.purringSounds.has(catId)) return;

        const head = visual.FindFirstChild("Head") as BasePart || visual.PrimaryPart;
        if (!head) return;

        // Check if sound already exists
        let sound = head.FindFirstChild("PurringSound") as Sound;
        if (!sound) {
            sound = new Instance("Sound");
            sound.Name = "PurringSound";
            // Using a placeholder sound ID - replace with actual purring sound asset ID
            // You can find cat purring sounds in the Roblox library or upload your own
            sound.SoundId = "rbxassetid://131961136"; // Placeholder: replace with actual cat purr sound
            sound.Volume = 0.4;
            sound.Looped = true;
            sound.Parent = head;
        }

        if (!sound.IsPlaying) {
            sound.Play();
            this.purringSounds.set(catId, sound);

            // Stop sound after 3 seconds (when purr action ends)
            task.delay(3, () => {
                this.StopPurringSound(catId);
            });
        }
    }

    public static CullDistantCats(playerPosition: Vector3) {
        this.catVisuals.forEach((visual, catId) => {
            if (visual.PrimaryPart) {
                const distance = visual.PrimaryPart.Position.sub(playerPosition).Magnitude;
                if (distance > 200) {
                    visual.Parent = undefined;
                } else {
                    visual.Parent = Workspace;
                }
            }
        });
    }
}
